File: 3453. Separate Squares I/go/solution_test.go
package main

import (
	"math/rand"
	"testing"
)

// Test functions for all solutions
func TestExample1_AllSolutions(t *testing.T) {
	squares := [][]int{{0, 0, 1}, {2, 2, 1}}
	expected := 1.00000
	epsilon := 1e-5

	result1 := (&Solution1BruteForce{}).SeparateSquares(squares)
	if abs(result1-expected) > epsilon {
		t.Errorf("Solution1: got %.5f, want %.5f", result1, expected)
	}

	result2 := (&Solution2BinarySearch{}).SeparateSquares(squares)
	if abs(result2-expected) > epsilon {
		t.Errorf("Solution2: got %.5f, want %.5f", result2, expected)
	}

	result3 := (&Solution3Mathematical{}).SeparateSquares(squares)
	if abs(result3-expected) > epsilon {
		t.Errorf("Solution3: got %.5f, want %.5f", result3, expected)
	}
}

func TestExample2_AllSolutions(t *testing.T) {
	squares := [][]int{{0, 0, 2}, {1, 1, 1}}
	expected := 1.16667
	epsilon := 1e-5

	result1 := (&Solution1BruteForce{}).SeparateSquares(squares)
	if abs(result1-expected) > epsilon {
		t.Errorf("Solution1: got %.5f, want %.5f", result1, expected)
	}

	result2 := (&Solution2BinarySearch{}).SeparateSquares(squares)
	if abs(result2-expected) > epsilon {
		t.Errorf("Solution2: got %.5f, want %.5f", result2, expected)
	}

	result3 := (&Solution3Mathematical{}).SeparateSquares(squares)
	if abs(result3-expected) > epsilon {
		t.Errorf("Solution3: got %.5f, want %.5f", result3, expected)
	}
}

func abs(x float64) float64 {
	if x < 0 {
		return -x
	}
	return x
}

// Benchmark data generation
func generateBenchmarkSquares(n int) [][]int {
	rand.Seed(42) // Fixed seed for reproducibility
	squares := make([][]int, n)

	for i := 0; i < n; i++ {
		squares[i] = []int{
			rand.Intn(1001),    // x: 0-1000
			rand.Intn(1001),    // y: 0-1000
			rand.Intn(100) + 1, // l: 1-100
		}
	}

	return squares
}

// ============================================================================
// Benchmarks: Solution 1 - Brute Force
// ============================================================================

func BenchmarkSolution1_10(b *testing.B) {
	squares := generateBenchmarkSquares(10)
	solution := &Solution1BruteForce{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution1_100(b *testing.B) {
	squares := generateBenchmarkSquares(100)
	solution := &Solution1BruteForce{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution1_1000(b *testing.B) {
	squares := generateBenchmarkSquares(1000)
	solution := &Solution1BruteForce{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution1_5000(b *testing.B) {
	squares := generateBenchmarkSquares(5000)
	solution := &Solution1BruteForce{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

// ============================================================================
// Benchmarks: Solution 2 - Binary Search (Recommended)
// ============================================================================

func BenchmarkSolution2_10(b *testing.B) {
	squares := generateBenchmarkSquares(10)
	solution := &Solution2BinarySearch{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution2_100(b *testing.B) {
	squares := generateBenchmarkSquares(100)
	solution := &Solution2BinarySearch{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution2_1000(b *testing.B) {
	squares := generateBenchmarkSquares(1000)
	solution := &Solution2BinarySearch{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution2_5000(b *testing.B) {
	squares := generateBenchmarkSquares(5000)
	solution := &Solution2BinarySearch{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

// ============================================================================
// Benchmarks: Solution 3 - Mathematical (Optimal)
// ============================================================================

func BenchmarkSolution3_10(b *testing.B) {
	squares := generateBenchmarkSquares(10)
	solution := &Solution3Mathematical{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution3_100(b *testing.B) {
	squares := generateBenchmarkSquares(100)
	solution := &Solution3Mathematical{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution3_1000(b *testing.B) {
	squares := generateBenchmarkSquares(1000)
	solution := &Solution3Mathematical{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

func BenchmarkSolution3_5000(b *testing.B) {
	squares := generateBenchmarkSquares(5000)
	solution := &Solution3Mathematical{}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		solution.SeparateSquares(squares)
	}
}

File: 3453. Separate Squares I/java/src/test/java/SolutionTest.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for all 3 solutions
 */
public class SolutionTest {
    
    private static final double EPSILON = 1e-5;
    
    private final Solution1BruteForce solution1 = new Solution1BruteForce();
    private final Solution2BinarySearch solution2 = new Solution2BinarySearch();
    private final Solution3Mathematical solution3 = new Solution3Mathematical();
    
    @Test
    public void testExample1_AllSolutions() {
        int[][] squares = {{0, 0, 1}, {2, 2, 1}};
        double expected = 1.00000;
        
        assertEquals(expected, solution1.separateSquares(squares), EPSILON, "Solution 1 failed");
        assertEquals(expected, solution2.separateSquares(squares), EPSILON, "Solution 2 failed");
        assertEquals(expected, solution3.separateSquares(squares), EPSILON, "Solution 3 failed");
    }
    
    @Test
    public void testExample2_AllSolutions() {
        int[][] squares = {{0, 0, 2}, {1, 1, 1}};
        double expected = 1.16667;
        
        assertEquals(expected, solution1.separateSquares(squares), EPSILON, "Solution 1 failed");
        assertEquals(expected, solution2.separateSquares(squares), EPSILON, "Solution 2 failed");
        assertEquals(expected, solution3.separateSquares(squares), EPSILON, "Solution 3 failed");
    }
    
    @Test
    public void testSingleSquare_AllSolutions() {
        int[][] squares = {{0, 0, 2}};
        double expected = 1.00000;
        
        assertEquals(expected, solution1.separateSquares(squares), EPSILON, "Solution 1 failed");
        assertEquals(expected, solution2.separateSquares(squares), EPSILON, "Solution 2 failed");
        assertEquals(expected, solution3.separateSquares(squares), EPSILON, "Solution 3 failed");
    }
    
    @Test
    public void testLargeSquare_AllSolutions() {
        int[][] squares = {{0, 0, 100}};
        double expected = 50.00000;
        
        assertEquals(expected, solution1.separateSquares(squares), EPSILON, "Solution 1 failed");
        assertEquals(expected, solution2.separateSquares(squares), EPSILON, "Solution 2 failed");
        assertEquals(expected, solution3.separateSquares(squares), EPSILON, "Solution 3 failed");
    }
    
    @Test
    public void testMultipleNonOverlapping_AllSolutions() {
        int[][] squares = {{0, 0, 1}, {0, 2, 1}, {0, 4, 1}};
        double expected = 2.50000; // Total area = 3, target = 1.5, line at y=2.5
        
        assertEquals(expected, solution1.separateSquares(squares), EPSILON, "Solution 1 failed");
        assertEquals(expected, solution2.separateSquares(squares), EPSILON, "Solution 2 failed");
        assertEquals(expected, solution3.separateSquares(squares), EPSILON, "Solution 3 failed");
    }
}
